README.md
Surface
=======

[![NPM version][npm-image]][npm-url] 
[![build status][travis-image]][travis-url] 
[![Test coverage][coveralls-image]][coveralls-url]
[![NPM Monthly Downloads][npm-download]][npm-url]
[![Dependencies][david-image]][david-url]
[![License][license-image]][license-url]
[![Tips][tips-image]][tips-url]

A tiny middleware of RESTful API for koa.

* Dependence on [koa-ovenware](https://github.com/zedgu/koa-ovenware)(koa-router).
* Support both JSON and XML format.
* Support customize response fields.
* Support global authentication.
* Add OPTIONS route for access control(CORS) automatically
* Write a controller and get all route pattern you want.
* Compatible with other middlewares including view renders.

Install
-------
```sh
npm install surface --save
```

Simple Usage
------------
####Require...
```js
var surface = require('surface');
```

####Config...
```js
surface(app);
```

####Controller file
Default path of controllers: ./lib/controllers/

in index.js:
```js
exports.index = function *() {
  this.body = 'hello koa';
};
```
Checkout the [examples](https://github.com/zedgu/surface/tree/master/examples).

####Response body
Request the root of the app, for example: http://localhost:3000/, will be:

#####in JSON
```json
{
  "request": "/",
  "code": 200,
  "message": "OK",
  "data": "hello koa"
}
```

#####in XML
```xml
<?xml version="1.0" encoding="UTF-8"?>
<response>
  <request>/</request>
  <code>200</code>
  <message>OK</message>
  <data>hello koa</data>
</response>
```

Conventions
-----------

####Action Mapping
```
route           http method    function of ctrl
:resource       get            index
:resource       post           create
:resource/:id   get            get
:resource/:id   put            update
:resource/:id   del            del
```
All routes can be customized by setting, see [Default values](#default-values); and also can be changed by controller api singly, see [APIs - Routes](#routes).

####Resource
Resource name will be the file name of the controller, if there is no alias set for the controller, see [APIs - Alias](#alias).

APIs
----
####Deprecated
```js
this.wrap // since 0.6.0

options.totally // since 0.6.0
```
####Options
```js
surface(app[, options])
```
`options` see [Default values](#default-values)
####Options.authenticate
To register a global authentication function.
####Options.deny
To set a function to handle the failing authentication.

**`authenticate` and `deny` have to be `Generator Function`.**
```js
suface(app, {
  /**
   * Global authentication
   * @return {Boolean}
   */
  authenticate: function *() {
    // do something and return true for authenticated, false for denied.
    // this === ctx
  },
  deny: function *() {
    // this.body...
    // this === ctx
  },
  authenticatePattern: /^\/api/i // default to the same as prefixPattern
});
```

####Controller APIs
#####Alias
Set alias for the controller.

```js
exports.alias = 'name_you_want';
```

#####Routes
Set routes for the controller.

```js
exports.routes = {
  entry: {
    method: 'get',
    path: '/index'
  }
};
```
#####Register route directly
To register route pattern directly, see [koa-router](https://github.com/alexmingoia/koa-router#routerverbname-path-middleware-middleware).

```js
app.register('http method', 'name of this route', 'route url pattern', callback);
```

#####Skip
Set true to not format by surface.

```js
ctx.skip_surface = true;
```

#####Model
Get model object.

```js
/**
 * get model object by given controller file name
 *
 * @param   {String}   modelName   optional, undefined for the model has
 *                                 the the same name as this controller
 * @return  {Object}               model object
 */
ctx.model([modelName])
exports.model([modelName])
```

for exmample:

```js
exports.get = function *() {
  this.model('abc'); // this === ctx
};
// or
exports.todo = function() {
  this.model(); // this === exports
};
```

#####Ctrl
Get controller object.

```js
/**
 * get ctrl object by given controller file name
 *
 * @param   {String}   ctrlName   optional, undefined for self
 * @return  {Object}              ctrl object
 */
ctx.ctrl([ctrlName])
exports.ctrl([ctrlName])
```

for exmample:

```js
exports.get = function *() {
  this.ctrl(); // this === ctx
  // => return this exports
};
// or
exports.todo = function() {
  this.ctrl('abc'); // this === exports
};
```

#####Format
Get the specifying format
- by query parameter
- by header `Accept`
- by default setting via options.format

> parmeter > Accept > options

Global configuration
--------------------
####Default values
```js
{
  root: './lib',        // root dir
  ctrl: 'controllers',  // controllers dir
  model: 'models'       // model dir
  format: 'json',       // format by default
  prefix: false,        // true,  only format the route match the prefixPattern;
                        // false, format all routes;
                        // String / RegExp, as the prefix of ALL routes.
                        // When `prefix` is a string / regexp and `options.prefixPattern` is not given, options.prefixPattern =  `new RegExp(prefix)` / `prefix`.
  prefixPattern: /^\/api\/v?\d{1,3}(\.\d{1,3}){0,2}/i,
                        // Only format the route match this pattern. Default to (not setting prefix and prefixPattern by `options`):
                        // /api/v1.1.1/**
                        // /api/0.0.1/**
                        // /api/1/**
  nosniff: true,        // X-Content-Type-Options
                        // see http://msdn.microsoft.com/library/ie/gg622941(v=vs.85).aspx
  options: 'Accept,Content-Type',
                        // false, not add OPTIONS routes for crossing domain requests automatically;
                        // String, to set Access-Control-Allow-Headers;
                        // see https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
  origin: '*',          // false, not add Access-Control-Allow-Origin header automatically;
                        // String, as value of Access-Control-Allow-Origin for all routes;
  authenticate: false,
  fields: {
    path: 'request',    // request url
    status: 'code',     // http status code
    message: 'msg',     // http status message
    data: 'data'        // real data
  },
  aliases: {
    'index': ''
  },
  routes: {
    'index': {
      method: 'get',
      path: ''
    },
    'create': {
      method: 'post',
      path: ''
    },
    'get': {
      method: 'get',
      path: '/:id'
    },
    'update': {
      method: 'put',
      path: '/:id'
    },
    'del': {
      method: 'delete',
      path: '/:id'
    }
  }
}
```


TODO
----
- API Docs

License
-------
MIT

[npm-image]: https://img.shields.io/npm/v/surface.svg?style=flat-square
[npm-url]: https://npmjs.org/package/surface
[travis-image]: https://img.shields.io/travis/zedgu/surface.svg?style=flat-square
[travis-url]: https://travis-ci.org/zedgu/surface
[coveralls-image]: https://img.shields.io/coveralls/zedgu/surface.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/zedgu/surface?branch=master
[david-image]: http://img.shields.io/david/zedgu/surface.svg?style=flat-square
[david-url]: https://david-dm.org/zedgu/surface
[npm-status]: https://nodei.co/npm/surface.png?downloads=true
[npm-status-url]: https://nodei.co/npm/surface/
[license-image]: http://img.shields.io/npm/l/surface.svg?style=flat-square
[license-url]: https://github.com/zedgu/surface/blob/master/LICENSE
[npm-download]: http://img.shields.io/npm/dm/surface.svg?style=flat-square
[tips-image]: http://img.shields.io/gittip/zedgu.svg?style=flat-square
[tips-url]: https://www.gittip.com/zedgu/


HISTORY.md
0.8.1 / 2016-06-02
==================
- [CHANGE] NO MORE `this.body = ''` for OPTIONS.
- [UPDATE] koa-ovenware to v0.3.1.

0.8.0 / 2016-02-02
==================
- [UPDATE] update all dependencies.
- [FIX] koa-router 5.x problems.

0.7.13 / 2016-01-22
==================
- [UPDATE] update all dependencies

0.7.12 / 2015-06-08
==================
- [FIX] update koa-ovenware to v0.2.0 to fix the windows issues.

0.7.11 / 2015-03-24
==================
- [ENHANCE] use koa-router OPTIONS handler to set Allow value.
- [FIX] koa-router 4.x problems.

0.7.10 / 2015-02-25
==================
- [FIX] some stupid bugs.

0.7.8 / 2015-02-09
==================
- [ADD] add Access-Control headers for crossing domain automatically.

0.7.7 / 2015-02-03
==================
- [ENHANCE] new code structure (use koa-ovenware for auto loading).
- [ADD] prefix could be a RegExp.
- [FIX] default format setting will really available by default (when there's no Accept head).
- [FIX] some bugs.

0.7.6 / 2014-12-10
==================
- [FIX] custom HTTP code

0.7.5 / 2014-10-16
==================
- [FIX] del route - `DELETE`
- [FIX] about https://github.com/koajs/koa/pull/353

0.7.4 / 2014-10-02
==================

- [REMOVE] No longer support `ctx.body = ''` => 204 No Content
  - use `ctx.body = null` instead
- [REMOVE] API - this.statusMessage
  - use `ctx.res.statusCode` and `ctx.res.statusMessage` (node native API) instead

0.7.3 / 2014-09-25
==================

- [ADD] default prefixPattern = new RegExp(prefix), if prefix is string.

0.7.2 / 2014-09-25
==================

- [ADD] API - this.statusMessage

0.7.1 / 2014-09-24
==================

- [ADD] prefix for all routes
- [FIX] #1

lib/surface.js
/**
 * Surface
 * Copyright 2014 Zed Gu
 * License: MIT
 */
'use strict';

/**
 * Dependencies
 */
var path = require('path');
var util = require('util');
var Kow = require('koa-ovenware');
var debug = require('debug')('surface:debug');

var render = require('./render');
var context = require('./context');

/**
 * Global
 */
var defaultSetting = {
  root: './lib',
  prefix: false,
  prefixPattern: /^\/api\/v?\d{1,3}(\.\d{1,3}){0,2}/i,
  ctrl: 'controllers',
  model: 'models',
  format: 'json',
  nosniff: true,
  authenticate: false,
  options: 'Accept,Content-Type',
  origin: '*',
  routes: {
    'index': {
      method: 'get',
      path: ''
    },
    'create': {
      method: 'post',
      path: ''
    },
    'get': {
      method: 'get',
      path: '/:id'
    },
    'update': {
      method: 'put',
      path: '/:id'
    },
    'del': {
      method: 'delete',
      path: '/:id'
    }
  },
  aliases: {
    'index': ''
  },
  fields: {
    path: 'request',
    status: 'code',
    message: 'msg',
    data: 'data'
  }
};

/**
 * Expose
 */
module.exports = Surface;

/**
 * Surface Constructor.
 * @param {Object} app     koa().
 * @param {Object} options Config object.
 * @api public
 */
function Surface(app, options) {
  if (!(this instanceof Surface)) {
    return new Surface(app, options);
  }

  this.conf = setting(options);

  this.middleware(app);
  context(app, this);
}

var surface = Surface.prototype;

surface.load = function(app, options) {
  var kp = Kow.prototype;
  var conf = this.conf;

  kp.preprocess = function(fn, path) {
    function *authFn(next) {
      if (yield conf.authenticate.bind(this)()) {
        yield fn.bind(this)();
      } else {
        yield conf.deny.bind(this)();
      }
      yield next;
    }

    function *Fn(next) {
      yield fn.bind(this)();
      yield next;
    }

    if (conf.authenticate) {
      if (conf.authenticatePattern && !conf.authenticatePattern.test(path)) {
        return Fn;
      }
      return authFn;
    }
    return Fn;
  };

  var kow = Kow(app, options);
  this.kow = kow;
  this.ctrls = kow.ctrls;
  this.models = kow.models;
};

/**
 * middleware of Surface
 * @return {Generator} middleware for koa
 */
surface.middleware = function(app) {
  var conf = this.conf;
  var fields = conf.fields;

  app.use(function() {
    if (conf.options) {
      return function *SurfaceOPTIONS(next) {
        yield *next;
        if (this.method === 'OPTIONS') {
          this.set('Access-Control-Allow-Methods', this.res.getHeader('Allow') + ', OPTIONS');
          this.set('Access-Control-Allow-Headers', conf.options);
          this.remove('Allow');
        }
      };
    } else {
      return function *SurfaceOPTIONS(next) {
        yield *next;
        if (this.method === 'OPTIONS') {
          this.remove('Allow');
          this.status = 404;
        }
      };
    }
  }());

  if (conf.nosniff) {
    app.use(function *SurfaceNosniff(next) {
      yield *next;
      this.set('X-Content-Type-Options', 'nosniff');
    });
  }

  if (conf.origin) {
    app.use(function *SurfaceOrigin(next) {
      yield *next;
      this.set('Access-Control-Allow-Origin', conf.origin);
    });
  }

  this.load(app, this.conf);

  var router = this.kow.router || app.router;
  // app.router for koa-router 4.x
  // this.kow.router for koa-router 5.x
  if (router.allowedMethods) {
    app.use(router.allowedMethods());
  }

  app.use(function *Surface(next) {
    yield next;

    if (this.skip_surface) {
      debug('%s : \nNot format by Surface via skip API', this.path);
      return;
    }

    if (conf.prefix && !conf.prefixPattern.test(this.path)) {
      debug('%s : \nNot format by Surface via prefix setting: %s', this.url, conf.prefixPattern);
      return;
    }

    var status = this.res.statusCode;
    if (status === 204 || status === 205 || status === 304) {
      return;
    }

    /**
     * ctx.toJSON().response.string for <= koa@0.12.2,
     * see https://github.com/koajs/koa/pull/353
     * and https://github.com/koajs/koa/commit/eb443d1bee748b3429d5308757573ec083e7e899#diff-d86a59ede7d999db4b7bc43cb25a1c11R477
     */
    var msg = this.res.statusMessage || this.toJSON().response.string || this.toJSON().response.message;
    var format = render.checkFormat.bind(this)(conf.format);

    this.body = render.format(format, this.path, status, msg, this.body, fields);
    this.type = format;
    this.res.statusCode = status;
  });
};

/**
 * Set conf
 * @param  {Object} options Options objest
 * @return {Object}         Formated conf object
 * @api private
 */
function setting(options) {
  var conf = {};

  if (options === null || typeof options !== 'object') {
    options = {};
  }

  for (var key in defaultSetting) {
    conf[key] = options[key] === undefined ? defaultSetting[key] : options[key];
  }

  conf.format = defaultFormat(['json', 'xml'].indexOf(conf.format.toLowerCase()));

  var prefix = conf.prefix;
  if ((typeof prefix === 'string' || util.isRegExp(prefix)) && !options.prefixPattern) {
    conf.prefixPattern = util.isRegExp(prefix) ? prefix : new RegExp(prefix);
  }

  if (conf.authenticate) {
    conf.deny = options.deny;
    conf.authenticatePattern = options.authenticatePattern || conf.prefixPattern;

    if (conf.authenticate.constructor.name !== 'GeneratorFunction') {
      console.error('authenticate should be a GeneratorFunction');
      conf.authenticate = false;
    }
    if (conf.deny.constructor.name !== 'GeneratorFunction') {
      console.error('deny should be a GeneratorFunction');
    }
  }

  return conf;
}

function defaultFormat(index) {
  var d = [['json', 'xml'], ['xml', 'json']];
  index = index === -1 ? 0 : index;

  return d[index];
}
